### PR TITLE
LLVM 4.0 rc2

### PR DESCRIPTION
**Pull request by dtzWill**
###### Motivation for this change

LLVM 4.0 is coming out soon, see https://github.com/NixOS/nixpkgs/issues/22630.

This adds LLVM 4.0 rc2 as a step towards support for LLVM 4.0 once it's released.

LLD is also included as it's now included in the LLVM release testing process.

###### Things done

- [x] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [x] macOS
   - [x] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: pkgs/development/compilers/llvm/4.0/libc++/default.nix [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+    unpackFile ${libcxxabi.src}
+    export LIBCXXABI_INCLUDE_DIR="$PWD/$(ls -d libcxxabi-${version}*)/include"
+  '';
+
+  preConfigure = ''
+    # Get headers from the cxxabi source so we can see private headers not installed by the cxxabi package
+    cmakeFlagsArray=($cmakeFlagsArray -DLIBCXX_CXX_ABI_INCLUDE_PATHS="$LIBCXXABI_INCLUDE_DIR")
+  '';
+
+  patches = lib.optional stdenv.isDarwin ./darwin.patch;
```
Comment (LnL7) [inline pkgs/development/compilers/llvm/4.0/libc++/default.nix]: this patch doesn't apply

File: pkgs/development/compilers/llvm/4.0/lldb.nix [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    # Fix up various paths that assume llvm and clang are installed in the same place
+    sed -i 's,".*ClangConfig.cmake","${clang-unwrapped}/lib/cmake/clang/ClangConfig.cmake",' \
+      cmake/modules/LLDBStandalone.cmake
+    sed -i 's,".*tools/clang/include","${clang-unwrapped}/include",' \
+      cmake/modules/LLDBStandalone.cmake
+    sed -i 's,"$.LLVM_LIBRARY_DIR.",${llvm}/lib ${clang-unwrapped}/lib,' \
+      cmake/modules/LLDBStandalone.cmake
+  '';
+
+  buildInputs = [ cmake python which swig ncurses zlib libedit llvm ];
```
Comment (LnL7) [inline pkgs/development/compilers/llvm/4.0/lldb.nix]: `libxml2` is missing for darwin https://gist.github.com/LnL7/af9d49197e2f26679ce76de2db20151b

**Comment by vcunat** [comment]
I assume this will wait until the final release, as it should happen within days?

File: pkgs/development/compilers/llvm/4.0/lldb.nix (hunk 2 above)
Comment (dtzWill) [inline pkgs/development/compilers/llvm/4.0/lldb.nix]: Whoops, my mistake.  Will be fixed in next push.

File: pkgs/development/compilers/llvm/4.0/libc++/default.nix (hunk 1 above)
Comment (dtzWill) [inline pkgs/development/compilers/llvm/4.0/libc++/default.nix]: [Here's the patch in question, as used for 3.9](https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/compilers/llvm/3.9/libc%2B%2B/darwin.patch)

Hmm.  While looking at the patch to understand it well enough to 'port' it, I noticed a few things:

1) The patch seems to be very close to `substituteInPlace lib/CMakeLists.txt --replace /usr/lib/libc++ '${LIBCXX_LIBCXXABI_LIB_PATH}/libc++'`.  Actually I just checked and on 3.9 at least it appears they're equivalent.

2) One of the paths changed with this patch is originally `/usr/lib/libc++.1.dylib` and afterward `${LIBCXX_LIBCXXABI_LIB_PATH}/libc++.1.dylib`.  This is used for the "install_name".  This seems odd since the currently-building libc++ library will never actually be put in that location (note the libc++abi library path is used).  I'm not sure what "install_name" is used for, but it seems like that this should instead be the path where libc++ will be after installation.  Is this a problem and if so is it worth addressing?

3) The substituteInPlace command above includes the "libc++" suffix duplicated in the patterns in order to avoid changing the path `/usr/lib/libSystem.B.dylib` [which is added to the linker flags on 10.6 apparently](https://github.com/llvm-mirror/libcxx/blob/release_39/lib/CMakeLists.txt#L116).  Is this the intended behavior, and if not should this be changed to a particular libSystem from some other place?

Long story short, I'm thinking drop the patch and just use substituteInPlace and call it good (patch is neater but substituteInPlace is clearer about what is being done)-- the question is whether the paths for "install_name" and libSystem should be changed or not.  I'll start by trying to preserve the 3.9 behavior but would like this addressed/decided before merging this, if possible.

Comment (dtzWill) [inline pkgs/development/compilers/llvm/4.0/libc++/default.nix]: Pushed commit that tries the substituteInPlace approach, let me know if that works for you-- looking at [the 4.0 version of this file](https://github.com/llvm-mirror/libcxx/blob/release_40/lib/CMakeLists.txt#L131) I think this will accomplish what the patch did previously.

As a minor note: at least on Linux it seems the /usr/lib/libSystem path gets rewritten to /var/empty/libSystem which definitely seems wrong.  Or maybe it doesn't matter :).

**Comment by dtzWill** [comment]
@vcunat I think it depends on whether they're doing an rc3 or not (which I think would likely delay things).

That said, I'm excited for 4.0 support (for example, my Souper update PR is blocking on this) and since this doesn't change the default llvmPackages I don't see any particular reason to not merge it once it's working on supported platforms.

Unless that seems like a bad idea to you/folks, LMK if you have any opinions/thoughts on the matter one way or the other :).

**Comment by vcunat** [comment]
llvm.org still states release scheduled at Feb 21.  It's no kind of a promise AFAIK, but I haven't seen any indication the real release date would be far.

I see no harm in merging it as RC2, except that this is a bit confusing:
```
nix-repl> llvmPackages_40.llvm.name
"llvm-4.0.0"
```

**Comment by dtzWill** [comment]
Okay, sounds good.

Fixing the name seems like a good idea regardless, so I've gone ahead and added that.

I asked over on #llvm (OFTC) and it seems they are likely to do another RC but nothing official (as you correctly observed) so dunno.

Until the Darwin build works it's a bit of a moot point, though! :grin:

**Comment by dtzWill** [comment]
Rebased to cleanup a bit of history.  Disabled LLDB on Darwin since it looks like fixing that will be its own project.

**Comment by LnL7** [comment]
I just built this on darwin 🎉

```
{ pkgs ? import nixpkgs {}, nixpkgs ? ./. }:

let

  inherit (pkgs) callPackage stdenvAdapters;

  hello = callPackage ./pkgs/applications/misc/hello {
    stdenv = stdenvAdapters.overrideCC pkgs.stdenv pkgs.llvmPackages_40.clang;
  };

in
  hello
```

**Review by LnL7 — APPROVED** [review]

**Comment by dtzWill** [comment]
Ping! If possible it'd be great to have this merged (or 4.0 final once it's ready) in time for the 17.03 branch O:).

**Comment by fpletz** [comment]
How did I merge this?! I didn't even have this PR open in my browser?! I merged other PRs...

**Comment by vcunat** [comment]
I did merge it.

**Comment by vcunat** [comment]
```
commit + d444aa4ade58fc4196dbcaba25e134f11736c514
Merge: 6992824ec10 e55b829c692
Author:     Vladimír Čunát <vcunat@gmail.com>
AuthorDate: Sun Feb 26 09:37:31 2017 +0100
Commit:     Vladimír Čunát <vcunat@gmail.com>
CommitDate: Sun Feb 26 10:19:13 2017 +0100

    Merge #22970: add 4.0-rc2 version of llvmPackages
```

And I pushed it to master.  It's probably another GitHub bug.  I'll ping their support.

**Comment by vcunat** [comment]
Back to llvm: now I see [versioning scheme has changed](http://lists.llvm.org/pipermail/llvm-dev/2016-December/107805.html), similarly to gcc, so we should perhaps call the attributes `*_4` instead of `*_40`.  Any thoughts?

**Comment by andrewrk** [comment]
https://github.com/NixOS/nixpkgs/pull/28410
